win32ss/user/user32/windows/cursoricon.c

```diff
@@ -1,17 +1,266 @@
 /*
- * PROJECT:         ReactOS user32.dll
- * COPYRIGHT:       GPL - See COPYING in the top level directory
- * FILE:            win32ss/user/user32/windows/cursoricon.c
- * PURPOSE:         cursor and icons implementation
- * PROGRAMMER:      Jérôme Gardou (jerome.gardou@reactos.org)
+ * PROJECT:     ReactOS user32.dll
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     cursor and icons implementation
+ * COPYRIGHT:   Copyright Jérôme Gardou <jerome.gardou@reactos.org>
+ *              Copyright 2022-2025 Doug Lyons <douglyons@douglyons.com>
+ *              Copyright 2025 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
  */
 
 #include <user32.h>
+#include "png.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(cursor);
 WINE_DECLARE_DEBUG_CHANNEL(icon);
 //WINE_DECLARE_DEBUG_CHANNEL(resource);
 
+#include <pshpack1.h>
+typedef struct _CURSORICONFILEDIRENTRY
+{
+    BYTE bWidth;
+    BYTE bHeight;
+    BYTE bColorCount;
+    BYTE bReserved;
+    union
+    {
+        WORD wPlanes; /* For icons */
+        WORD xHotspot; /* For cursors */
+    };
+    union
+    {
+        WORD wBitCount; /* For icons */
+        WORD yHotspot; /* For cursors */
+    };
+    DWORD dwDIBSize;
+    DWORD dwDIBOffset;
+} CURSORICONFILEDIRENTRY;
+
+typedef struct _CURSORICONFILEDIR
+{
+    WORD idReserved;
+    WORD idType;
+    WORD idCount;
+    CURSORICONFILEDIRENTRY idEntries[1];
+} CURSORICONFILEDIR;
+#include <poppack.h>
+
+#define PNG_CHECK_SIG_SIZE 8 /* Check signature size */
+
+/* libpng helpers */
+typedef struct _PNG_READER_STATE
+{
+    png_bytep buffer;
+    size_t bufSize;
+    size_t currentPos;
+} PNG_READER_STATE;
+
+/* This function is used for reading png data from memory */
+static VOID
+ReadMemoryPng(
+    _Inout_ png_structp png_ptr,
+    _Out_ png_bytep data,
+    _In_ size_t length)
+{
+    PNG_READER_STATE *state = png_get_io_ptr(png_ptr);
+    if ((state->currentPos + length) > state->bufSize)
+    {
+        ERR("png read error\n");
+        png_error(png_ptr, "read error in ReadMemoryPng");
+        return;
+    }
+    RtlCopyMemory(data, state->buffer + state->currentPos, length);
+    state->currentPos += length;
+}
+
+/* libpng.dll is delay-loaded. If no libpng.dll exists, we have to avoid exception */
+static BOOL
+LibPngExists(VOID)
```

Comment:
Ideally this could be handled a bit more properly with delay-load hooks, but let's keep it this way for the purpose of this PR.